Detect geo of server.

sofar opened the pull request:
Using geolite2, we attempt to detect the continent of the server
and store the 'continent code' in the serverlist. This is reasonably
broad enough to help players in the client find truly "nearby" server.

The client will have to be changed to select a continent to change
orderign or filtering.

Minetest-One commented:
Good , better then nothing,
best is still client might make own ping, as solved ALL problems.
I will see how it will work, as have no idea how it will be implemented  ...
as there is always a gap between wish and reality :-)

sofar commented:
I think I've mentioned it before, but I think it would be too costly for each MT client to ping each MT server. This would effectively mean millions of ping packets on a time scale well under an hour. I do not think that that is a scalable solution, and I therefore do not support such an implementation.

nerzhul commented on server.py:
```diff
@@ -270,13 +271,24 @@ def asyncFinishThread(server):
+			reader = geolite2.reader()
+			geo = reader.get(addr)
+			server["geo_continent"] = geo["continent"]["code"]
+	except:
```
ugly exception catch. Please catch exception properly not all exceptions

sofar replied:
All I can think of is to take `for addr in addresses:` out, but that's it.

This is and probably should be "optional" where if it doesn't work, we shouldn't crash the server thread. If people don't have the right DB files in place, it should still work.

Even if this is long in production, clients still should be able to handle servers in the list without geo hint, so this should therefore remain optional and not crashing the python stack, no matter the error. Note that the code no matter what will print out a log message indicating that there was a failure, so I don't feel that we should start handing exceptions with the depth that is suggested at https://github.com/maxmind/GeoIP2-python#database-reader-exceptions .

Or I just don't understand your suggestion... I do confess that Python is my #1 most hated and avoided language, I really don't like it and this is the most python code I've written in a afew months, and I'm convinced you and many others could do a better job with 3 minutes time.

nerzhul replied:
the problem i mention is the exception is too large, you should handle only geoip exception, not all exception which can be coding errors :)
handling it in the loop is desired :)

sfan5 commented on server.py:
```diff
@@ -270,13 +271,24 @@ def asyncFinishThread(server):
 			return
 
+	reader = geolite2.reader()
+	for addr in addresses:
```
I'd just run this on `server["ip"]` once, instead of running it on all the results of DNS query

sfan5 commented on server.py:
```diff
@@ -270,13 +271,24 @@ def asyncFinishThread(server):
+		except:
+			pass
+
+	if not server["geo_continent"]:
```
this will actually raise an exception if the key does not exist
you probably want `if "geo_continent" not in server:`

sfan5 commented on server.py:
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 import os, re, sys, json, time, socket
 from threading import Thread, RLock
+from geolite2 import geolite2
```
should be added to requirements.txt

sofar commented:
Updated:
- added requirements.txt line
- exception only handles the core geolite2 exception
- warnings are a bit more detailed
- still fairly fault tolerant as is

Seems a little bit cleaner now as well.

sfan5 commented:
Merged. 772fc29cb89ccd0052a64a8a9caf7cb4e8f481e9
